#2 Signal inputs

devversion opened the pull request:
Please refer to #53521 for information on signal inputs. This is the second PR towards completing input signals.

See individual commits.

devversion committed:
test(compiler-cli): additional type-check transform tests for signal inputs

This commit adds additional type-check transform tests for signal
inputs. These tests verify some of the problems with covariance,
contravariance and bivariance that we were suspecting to be problematic
if we would assign `InputSignal`'s directly to the type constructors.

devversion committed:
refactor(core): introduce runtime `InputSignal` implementation

This commit introduces the runtime `InputSignal` implementation.
Input initializers using `input` or `input.required` will result in
an instance of `InputSignal` to be created.

An input signal extends the signal primtive, with a couple of small
differences:

 - it's a readonly signal. There is no public `set` or `update`.
 - equality is non-configurable. As per CD semantics, the value is
   guaranteed to be different when the `property` instruction attempts
   to update an input signal.
 - we support a `transform` function, that allows transforming input
   values. The transform is called whenever the input is set. An
   alternative could have been to follow computed-semantics and call the
   transform upon accessing, if dirty.

In the future, we might change this to extend the computed reactive
node, so that we can support computed inputs that do not rely on
continious bound value assignments. See signal based components RFC.

devversion committed:
refactor(core): initial test code for `setInput` to work with input signals

At this point, we have the following pieces in place:

* the input signature is implemented
* the compiler properly parses and recognizes signal inputs
* the compiler supports type-checking of signal inputs
* input signal metadata is passed to partial output

This commit adds a naive runtime solution to distinguishing between
signal inputs and decorator inputs when the `property` instruction
invokes. This is not ideal and non-performant as we introduce additional
megamorphic reads for every property instruction invocation, or if we'd
use `instanceof`, introducing a hard dependency on `InputSignal` and
risking potentially slower detection.

This code exists purely for testing, to enable playing with input
signals in the playground. In a future commit, we will pass around the
input signal metadata at runtime and can perform highly optimized checks
to distinguish between signal or non-signal inputs- when assigning
values.

More information: https://docs.google.com/document/d/1FpnFruviKb6BFTQfMAP2AMEqEB0FI7z-3mT_qm7lzX8/edit#heading=h.oloxympe902x

devversion committed:
refactor(core): detect signal inputs at runtime using input flags

This commit introduces a new enum for capturing additional metadata
about inputs. Called `InputFlags`. These will be built up at compile
time and then propagated into the runtime logic, in a way that does
not require additional lookup dictionaries data structures, or
additional memory allocations for "common inputs" that do not have any flags.

The flags will incorporate information on whether an input is signal
based. This can then be used to avoid megamorphic accesses when such
input is set- as we'd not need to check the input field value. This also
avoids cases where an input signal may be used as initial value for an
input (as we'd not incorrectly detect the input as a signal input then).

The new metadata emit will be useful for incorporating additional
metadata for inputs, such as whether they are required etc (although
required inputs are a build-time only construct right now- but this is a
good illustration of why input flags can be useful). An alternative
could have been to have an additional boolean entry for signal inputs,
but allocating a number with more flexible input flags seems more future
proof and more reasonable andreadable.

More information on the megamorphic access when updating an input
signal
https://docs.google.com/document/d/1FpnFruviKb6BFTQfMAP2AMEqEB0FI7z-3mT_qm7lzX8/edit.

devversion committed:
refactor(core): type `EMPTY_OBJ` as `never` for improved type safety

Consider a snippet like:

```
const x = directiveDef.inputs || EMPTY_OBJ
```

this currently results in `x` being inferred as just `{}`- ending up
turning of potential future assignment checks. This surfaced in the
`DirectiveDefinition` -> `DirectiveDef` conversion.

Note: This has the effect that assigning `EMPTY_OBJ` to a field of
anything would _always_ pass. It's questionable if this rather impacts
type-safety in a more negative way. There seem to be trade-offs in both
ways... Maybe worth considering just using `{}` directly as fallbacks in
some places, and treating this as an unique symbol?!

https://www.typescriptlang.org/play?#code/MYewdgzgLgBAHgLhmApgNxQJxgXhgbwF8YBDCZdLAbgCgbRJYAHJfGmDmAGxC6SkwBXFABoahAD6CwAExQAzAJaoZuZIK5dS5EmACeteuGgxBapjAkT4VIA

devversion committed:
test(core): add type signature test for signal input API

Adds tests that allow us to ensure that the `input` API works as
expected and that resulting return types match our expectations- without
silently regressing in the future, or missing potential edge-cases.

Testing signatures is hard because of covariance and contravariance,
especially when it comes to the different semantics of `ReadT` and
`WriteT` of input signals. We enable reliable testing by validating the
`d.ts` of the "fake directive class". This ensures clear results,
compared to relying on e.g. type assertions that might
accidentally/silently pass due to covariance/contravariance or
biavariance in the type system.

devversion committed:
test(compiler-cli): add ngtsc test for new signal input API

This commit adds a final test for input signals, integrating all major
parts:

* type-checking
* compiler detection
* compiler emit
* API signature tests

devversion committed:
build: re-enable linker compliance tests

The linker compliance tests were disabled with a Babel update and
nobody realized for quite a while, via
https://github.com/angular/angular/pull/49914.

As we've came across this lost coverage, which also is quite
impactful as all libraries depend on linked output- I've took initiative
to debug the root cause as there was no follow-up.
https://github.com/angular/angular/issues/51647.

It turned out to be a highly complex issue that is non-trivial to fix,
but at least we should try to resurrect the significant portion of test
coverage by still running the linker tests- avoiding regressions, or
unexpected issues (like with defer being developed). We can work on
re-enabling and fixing source-maps separately.

Tracked via https://github.com/angular/angular/issues/51647.

devversion committed:
refactor(compiler-cli): reference `InputFlags` enum directly for full compiler output

Instead of computing the bit input flags at compile-time and inling
the final bit flag number, we will use the `InputFlags` enum directly.
This is a little more code in the compiler side, but will allow us to
have better debuggable development code, and also prevents problems
where runtime flag bitmasks differ from the compiler flag bitmasks.

This is in practice a noop for optimized applications as the enum values
would be inlined anyway. This matches existing compiler emit for e.g.
change detection strategy, or view encapsulation enums.

devversion committed:
test: fix linker compliance tests after not running for a while

The linker compliance tests did not run for a while. There were a couple
of new tests that were not passing as this wasn't flagged on CI. This commit fixes this.
Fortunately there was no problematic code that did indicate issues with linking.

In the follow-up commit, we fix the compliance test infrastructure to
re-enable linker testing..

One clear issue is still that the defer blocks are not handled properly
in linked output- hence making defer not actually "lazy" for compiled
libraries. This needs to be handled separately by the framework team.

devversion committed:
fix(core): `SignalNode` reactive node incorrectly exposing unset field

The `SignalNode` interface, describing the reactive node for a `Signal`,
seemingly exposes the `SIGNAL` symbol as a class member. This is not
true as the `SIGNAL` reactive node only exists on the getter function,
as a way to retrieve the signal underlying reactive node.

This commit fixes this, enabling improved type-safety later, in a
follow-up commit where `SIGNAL_NODE` can now be typed to match the
`SignalNode` interface (unlike now where it's typed as just `object`).

devversion committed:
refactor(core): expose `SIGNAL_NODE` to allow for advanced extensions of signals

For the implementation of input signals, we want to extend the signal
primitive. The basic methods exposed here are not suitable as we'd like
to store additional metadata on the reactive node, and also have a
custom getter (for required inputs and throwing).

To enable this, one small piece was missing. This commit exposes it and
also improves type safety, now that `SignalNode` is typed properly after
the previous commit.

JoostK commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -619,47 +618,103 @@ export function createTNode(
-      } else if (hostDirectiveAliasMap.hasOwnProperty(publicName)) {
-        addPropertyAlias(
-            propertyAliases, directiveIndex, hostDirectiveAliasMap[publicName], internalName);
+    const value = aliasMap[publicName];
```
`hasOwnProperty`

JoostK commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -619,47 +618,103 @@ export function createTNode(
+      continue;
+    }
+
+    bindingsResult = bindingsResult === null ? {} : bindingsResult;
```
nit:
```suggestion
    bindingsResult ??= {};
```

devversion replied:
Done (this was code from before- but good cleanup)

JoostK commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -619,47 +618,103 @@ export function createTNode(
+    bindingsResult = bindingsResult === null ? {} : bindingsResult;
+
+    let internalName: string;
+    let inputFlags: InputFlags = 0;
```
nit:
```suggestion
    let inputFlags = InputFlags.None;
```

JoostK commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -619,47 +618,103 @@ export function createTNode(
+    hostDirectiveAliasMap: HostDirectiveBindingMap|null): NodeOutputBindings|null;
+
+function captureNodeBindings<T>(
+    type: 'inputs'|'outputs', aliasMap: DirectiveDef<T>['inputs']|DirectiveDef<T>['outputs'],
```
Can `type` be an enum?

devversion replied:
The generic parameter does actually improve type safety because otherwise `inputs` previously were not typed properly and always ended up as `{}`. These changes improve type-safety. This was realized when I made structural changes to the internal data structures, while FW compiled without any errors!

Note that we already have two overload signatures. This is just the implementation. We could improve this a bit by having two full functions, but then duplication for the code is risky. A single function is the optimization here I think.

devversion replied:
Changing to an enum

JoostK commented on packages/core/src/render3/interfaces/node.ts:
````diff
@@ -880,29 +881,37 @@ export interface TProjectionNode extends TNode {
+ * ```
  */
-export type PropertyAliasValue = (number|string)[];
+export type NodeInputBindings = Record<string, (number|string|InputFlags)[]>;
````
The value type could be a tuple here

devversion replied:
Nice idea!

devversion replied:
Turned out this doesn't work as we need the values to be "repeatable" (the tuple can occur X times)

JoostK commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -930,11 +939,12 @@ export type InitialInputData = (InitialInputs|null)[];
  * e.g. ['role-min', 'minified-input', 'button']
  */
-export type InitialInputs = string[];
+export type InitialInputs = (string|InputFlags)[];
```
The value type could be a tuple here

devversion replied:
👍

JoostK commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +470,36 @@ function assertDomElement(value: any) {
+  for (const key in inputs) {
+    const value = inputs[key];
+
+    if (value === undefined) {
```
`hasOwnProperty`

JoostK commented on packages/core/test/authoring/type_tester.ts:
```diff
@@ -0,0 +1,61 @@
+  let failing = false;
+
+  for (const member of testClazz.members) {
+    if (ts.isPropertyDeclaration(member) === false) {
```
nit:
```ts
    if (!ts.isPropertyDeclaration(member)) {
```

JoostK commented on packages/core/test/authoring/type_tester.ts:
```diff
@@ -0,0 +1,61 @@
+const containingDir = path.dirname(url.fileURLToPath(import.meta.url));
+const testDtsFile = path.join(containingDir, 'signal_input_signature_test.d.ts');
+
+async function main() {
```
💯

JoostK commented on packages/compiler-cli/test/ngtsc/authoring_spec.ts:
```diff
@@ -0,0 +1,192 @@
+    });
+
+
+    it('should handle a non-primitive initial values', () => {
```
```suggestion
    it('should handle non-primitive initial values', () => {
```

(or drop the plural `s` for `values`)

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/lifecycle_hooks/lifecycle_hooks_lifecycle_comp_def.js:
```diff
@@ -1,7 +1,7 @@
   type: LifecycleComp,
   selectors: [["lifecycle-comp"]],
-  inputs: {nameMin: [0, "name", "nameMin"]},
+  inputs: {nameMin: [$r3$.ɵɵInputFlags.None, "name", "nameMin"]},
```
I wouldn't personally have gone this route. The compliance tests support `__` prefix and suffix to have readable bitflags (e.g. `__AttributeMarker.Template__`) in expectations and emitting just the integer reduces reliance on a build step to inline the values.

devversion replied:
We can totally drop that commit if we want. My thinking was such inlining of this is trivial, and we also have more debuggable code. There are many places where we rely on build optimization in the framework

JoostK reviewed (commented):
Nice work! 👏

devversion commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -204,14 +206,29 @@ export function conditionallyCreateDirectiveBindingLiteral(
-        if (value.transformFunction != null) {
-          expressionKeys.push(value.transformFunction);
+      // Build up input flags
+      let flags = 0;
```
I intentionally did it that way to somehow indicate that these are bit flags, and a None flag would kind of seem confusing, conceptually. TS should still check this properly IIRC

alxhub commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/lifecycle_hooks/lifecycle_hooks_lifecycle_comp_def.js:
```diff
@@ -1,7 +1,7 @@
   type: LifecycleComp,
   selectors: [["lifecycle-comp"]],
-  inputs: {nameMin: ["name", "nameMin"]},
+  inputs: {nameMin: [0, "name", "nameMin"]},
```
I suppose we _could_ omit the `0` for inputs without flags... it's probably not worth it though.

devversion replied:
Yeah, I did have that implemented initially, but it didn't feel worth doing as such structure would only exist for non-signal aliased inputs, or those with transforms. The structure never was fully optimal.

e.g. if there was a transform, we captured an alias as well (even though it would be the same as the "public name").

alxhub commented on packages/core/src/authoring/input_signal_node.ts:
```diff
@@ -0,0 +1,49 @@
+    applyValueToInputSignal<ReadT, WriteT>(
+        node: InputSignalNode<ReadT, WriteT>, value: ReadT|WriteT): void {
+      const finalValue =
+          node.transformFn !== undefined ? node.transformFn(value as WriteT) : (value as ReadT);
```
Seeing this change made me think of a problem with the logic here. For normal inputs, `ngOnChanges` observes the post-transformed value (after it's transformed in `writeToDirectiveInput`). This makes sense, as the `ngOnChanges` hook runs on the consumer/"read" side of the binding.

This doesn't happen properly for signal inputs, as the transform isn't applied in `writeToDirectiveInput` but rather internally within the `InputSignal`. We should probably move the transform to happen at the same time as normal inputs, rather than make it an internal part of the input signal impl.

devversion replied:
Fixed, as we discussed offline as well

devversion commented on packages/core/src/authoring/input.ts:
```diff
@@ -61,9 +61,9 @@ export function inputRequiredFunction<ReadT>(opts?: InputOptionsWithoutTransform
+export function inputRequiredFunction<ReadT, WriteT>(opts?: InputOptions<ReadT, WriteT>):
     InputSignal<ReadT, WriteT> {
-  throw new Error('TODO');
+  return createInputSignal(undefined as never, true, opts);
```
I do actually like that idea. I think the separation is acceptable.

devversion commented on packages/core/src/authoring/input_signal.ts:
```diff
@@ -45,7 +49,45 @@ export const ɵINPUT_SIGNAL_BRAND_WRITE_TYPE = /* @__PURE__ */ Symbol();
+  const node: InputSignalNode<ReadT, WriteT> = Object.create(INPUT_SIGNAL_NODE);
+
+  if (options?.transform !== undefined) {
+    node.transformFn = options.transform;
```
Done, adding this as a base explicit field to `InputSignalNode`

JoostK replied:
I don't know if setting it on the prototype actually has the desired effect, as the object created from `Object.create(INPUT_SIGNAL)` still has various property transitions, it either has the field itself, or delegates to its prototype. So although the read always _finds_ a property, they are still two distinct properties.

devversion replied:
Confirmed this by exploring v8 execution perf logs. good point! also captured in https://docs.google.com/document/d/1FpnFruviKb6BFTQfMAP2AMEqEB0FI7z-3mT_qm7lzX8/edit?tab=t.0

devversion commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -204,14 +206,29 @@ export function conditionallyCreateDirectiveBindingLiteral(
+      // Build up input flags
+      let flags = 0;
+      if (value.isSignal) {
+        flags = flags | InputFlags.SignalBased;
```
(note this is obsolete with us using the actual enums - resolving)

devversion commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -204,14 +206,29 @@ export function conditionallyCreateDirectiveBindingLiteral(
+
+      // Inputs, compared to outputs, will track their declared name (for `ngOnChanges`), or support
+      // transform functions, or store flag information if there is any.
+      if (forInputs && (differentDeclaringName || hasTransform || flags !== 0)) {
```
(same as above)

devversion commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -68,13 +68,20 @@ export class ComponentFactoryResolver extends AbstractComponentFactoryResolver {
-    if (map.hasOwnProperty(nonMinified)) {
-      const minified = map[nonMinified];
-      array.push({propName: minified, templateName: nonMinified});
+    const value = map[nonMinified];
+
+    if (value === undefined) {
```
I think we need both now given the improved types where `ComponentDef.inputs` may not be set for all `keyof T`. A separate generic for inputs/output keys would probably be better, but that's a different story.

devversion commented on packages/core/src/render3/definition.ts:
```diff
@@ -425,45 +440,67 @@ export function ɵɵdefineNgModule<T>(def: {
  * {
  *  'propName1': 'minifiedPropName1',
  *  'publicName2': 'minifiedPropName2',
+ *  'inputSignal': ['minifiedInputSignal', InputFlags.SignalBased],
```
I will probably keep it as is- simply because those are totally different data structures and this is probably less confusing if it would not look like the `ComponentDef` data strucure.

devversion commented on packages/core/src/render3/definition.ts:
```diff
@@ -641,7 +678,7 @@ export const GENERATED_COMP_IDS = new Map<string, Type<unknown>>();
  * algorithm.
  */
-function getComponentId(componentDef: ComponentDef<unknown>): string {
+function getComponentId<T>(componentDef: ComponentDef<T>): string {
```
I think here probably not. I believe might be an artifact from some other types I was testing with. Removed

devversion replied:
Actually keeping for same reasons as in the comment below.

devversion commented on packages/core/src/render3/features/host_directives_feature.ts:
```diff
@@ -186,11 +186,11 @@ function validateHostDirective(
  */
-function validateMappings(
-    bindingType: 'input'|'output', def: DirectiveDef<unknown>,
+function validateMappings<T>(
+    bindingType: 'input'|'output', def: DirectiveDef<T>,
```
The generic is actually beneficial as `keyof unknown` would otherwise turn `def.inputs` and `def.outputs` into just `{}`, totally loosing out on type safet.

devversion commented on packages/core/src/render3/features/input_transforms_feature.ts:
```diff
@@ -17,7 +17,7 @@ import {DirectiveDef, InputTransformFunction} from '../interfaces/definition';
  * @codeGenApi
  */
-export function ɵɵInputTransformsFeature(definition: DirectiveDef<unknown>): void {
+export function ɵɵInputTransformsFeature<T>(definition: DirectiveDef<T>): void {
```
Keeping for type-safety. See comment above.

devversion commented on packages/core/src/render3/features/host_directives_feature.ts:
```diff
@@ -203,8 +203,7 @@ function validateMappings(
 
-      if (bindings.hasOwnProperty(remappedPublicName) && remappedPublicName !== publicName &&
-          bindings[remappedPublicName] !== publicName) {
+      if (bindings.hasOwnProperty(remappedPublicName) && remappedPublicName !== publicName) {
```
The minified name of the directive binding should not impact the error as far I understood, so I removed this condiiton. cc. @crisbeto

crisbeto replied:
Hmm it's been a while so I don't really remember. It seems to me like something I left in there just to be safe and since the extra bytes don't matter because it's in a dev-mode-only assertion. I do see that initially the `hasOwnProperty` [check wasn't there](https://github.com/crisbeto/angular/commit/50f8928d56bdbc1e65dfd5127ec00d8d934d2432#diff-c67a1318ee71d2a5c85c4897736fc7cc30375590000c3699119bfc509415c21bR161) and [was added later](https://github.com/crisbeto/angular/commit/8b44ba31701a1c1cf1ec92d2a26f9cf657f5408b#diff-c67a1318ee71d2a5c85c4897736fc7cc30375590000c3699119bfc509415c21bL207) so maybe whatever I was guarding against initially was covered already by the `bindings.hasOwnProperty(remappedPublicName)`.

devversion replied:
sounds like this change is fine then?

crisbeto replied:
Yeah let's leave it. As long as the tests pass we should be okay.

JeanMeche commented on packages/core/src/authoring/input_signal.ts:
```diff
@@ -45,7 +49,45 @@ export const ɵINPUT_SIGNAL_BRAND_WRITE_TYPE = /* @__PURE__ */ Symbol();
+    producerAccessed(node);
+
+    if (node.value === REQUIRED_UNSET_VALUE) {
+      throw new Error('Input is required, but no value set yet.');
```
Can we have a `RuntimeError` here  ?

devversion replied:
In the third PR we will have a commit that changes this + adds error guides for aio and adev

devversion replied:
(will add a TODO)

devversion commented on packages/core/src/render3/definition.ts:
```diff
@@ -425,45 +440,67 @@ export function ɵɵdefineNgModule<T>(def: {
+    obj: DirectiveDefinition<T>['inputs'], declaredInputs: Record<string, string>):
+    Record<keyof T, string|[minifiedName: string, flags: InputFlags]>;
+
+function invertObject<T>(
```
Done/

devversion committed:
fixup! test(compiler-cli): add ngtsc test for new signal input API

Fixes a typo in a unit test, reported in PR review.

devversion committed:
fixup! refactor(core): detect signal inputs at runtime using input flags

Reflect the new input flags for inputs having transforms

devversion committed:
fixup! test(core): add type signature test for signal input API

Simplify if check

devversion committed:
fixup! refactor(core): introduce runtime `InputSignal` implementation

Simplify required signal node creation as per feedback

devversion committed:
fixup! refactor(core): introduce runtime `InputSignal` implementation

Transforms are run manually, plus ensure monomorphic access.

devversion committed:
fixup! refactor(core): introduce runtime `InputSignal` implementation

Follow-up TODO for runtime error. Commit ready for 3rd PR

devversion committed:
fixup! refactor(core): detect signal inputs at runtime using input flags

Address review feedback

e98870c6a18f34cbfd00702411e7f47d625ad8e6

devversion committed:
fixup! refactor(core): detect signal inputs at runtime using input flags

Introduce new `HasTransform` flag that will be used

Fix transforms for input signals with ngOnChanges. The value needs to be transformed before

devversion committed:
fixup! refactor(compiler-cli): reference `InputFlags` enum directly for full compiler output

Update after new HasTransform flag

devversion commented:
@alxhub @JoostK Pushed all addressed feedback as fixup commits (in between commits needed for order of changes). All individual commits auto-squashed should pass in order.

JoostK commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -72,8 +72,11 @@ function toRefArray<T>(map: {[P in keyof T]?: string|[minifiedName: string, flag
   const array: {propName: string; templateName: string;}[] = [];
   for (let nonMinified in map) {
-    const value = map[nonMinified];
+    if (!map.hasOwnProperty(nonMinified)) {
```
With this in place, I think the `undefined` check could be dropped. Also noticing now that `nonMinified` could be `const`

devversion replied:
The field can be undefined given the type of `map` above (type-wise - in practice it never happens), but with me fixing that these fields are properly type-checked now- this has surfaced here and below. I think this is fine?

JoostK replied:
Oh, yeah that is a bit unfortunate. Fine to keep the check :+1:

devversion replied:
Updated the `nonMinified` while being at it + gave it a better name.

JoostK commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -482,8 +482,11 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
 
   for (const key in inputs) {
-    const value = inputs[key];
+    if (!inputs.hasOwnProperty(key)) {
```
Similarly here for the `undefined` check

devversion replied:
(let's discuss above)

JoostK commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -618,6 +618,12 @@ export function createTNode(
 }
 
+/** Mode for capturing node bindings. */
+enum CaptureNodeBindingMode {
```
Since this is a local enum, this can be safely made `const` without affecting local compilation mode.

devversion replied:
My understanding is that enum inlining should be something optimizers should do. e.g. ESBuild, terser or Closure- and it was a "mistake" that TypeScript actually added this "feature". Happy to add this, but don't think it matters

JeanMeche replied:
Please correct me if I'm wrong somewhere: 

Right now the optimizers aren't capable of doing that, right ?
And since this is code that is shipped into the main bundle, let's be aggresive on the bundle impact. 

Also, if I understand Daniel Rosenwasser ([here](https://github.com/facebook/create-react-app/pull/4837#issuecomment-430107471)) correctly, this is one of the usecases where `const enum` makes the most sense.

devversion replied:
Looks like it would only be JSCompiler that can optimize these further, but even then not as good as with const enums. There is a transform in the Angular CLI that I thought would help here- but doesn't look like it. I will try adding const enum (but need to verify this is still allowed in g3)

JoostK replied:
Hmm, that's a bit unexpected; esbuild should be able to inline enum values. Perhaps there are some limitations w.r.t. cross-module/cross-chunk inlining or further tree-shaking capabilities, but I'd expect it to inline+tree-shake for this local case. I was triggered because `CaptureNodeBindingMode` did show up in the symbol goldens.

JoostK commented on packages/core/src/render3/instructions/write_to_directive_input.ts:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    if ((flags & InputFlags.HasTransform) !== 0) {
+      if (inputSignalNode !== null) {
+        value = inputSignalNode.transformFn!(value);
+      } else {
+        value = def.inputTransforms![privateName]!.call(instance, value);
+      }
```
⏩ Maybe something to consider for a future update could be to store the transform consistently in `def.inputTransforms`, now that it's not driven from within the input node itself.

devversion replied:
I've chatted with Alex about this, but my feeling was that this would complicate things unnecessarily because we'd then also need to setup the transform runtime feature for these components

JoostK replied:
I hadn't realised those were driven by a feature; then it makes perfect sense to store it in the input signal instead of out-of-band in the component definition.

devversion committed:
fixup! test: fix linker compliance tests after not running for a while

Disable new defer complaince tests to run with the linker. We disabled
most of these already, but new ones were introduced 3 days ago. This
commit disables those as well because defer is not yet handled properly
with partial compilation output (This is a known & planned task).

The tests started failing because in this PR we fixed that the linker
tests were not running for quite a while..

JoostK commented on packages/core/src/authoring/input_signal_node.ts:
```diff
@@ -0,0 +1,48 @@
+export const INPUT_SIGNAL_NODE: InputSignalNode<unknown, unknown> = /* @__PURE__ */ (() => {
+  return {
+    ...SIGNAL_NODE,
+    transformFn: undefined,
```
Forgot to mention, but I suppose this can be removed now (albeit with after changing the type to `Omt<..., 'transformFn'>`

devversion replied:
Yeah, considered doing that, but remembered other signal code following a similar pattern. See: https://github.com/angular/angular/blob/main/packages/core/primitives/signals/src/signal.ts#L99

pkozlowski-opensource commented on packages/core/src/util/empty.ts:
```diff
@@ -14,7 +14,7 @@ import {initNgDevMode} from './ng_dev_mode';
  */
 
-export const EMPTY_OBJ: {} = {};
+export const EMPTY_OBJ: never = {} as never;
```
👍

pkozlowski-opensource commented on packages/core/test/authoring/signal_input_signature_test.ts:
```diff
@@ -0,0 +1,98 @@
+// import preserved to simplify `.d.ts` emit and simplify the `type_tester` logic.
+import {InputSignal} from '../../src/authoring/input_signal';
+
+export class InputSignatureTest {
```
Those tests are super-useful! 👍

pkozlowski-opensource commented on packages/core/src/render3/instructions/write_to_directive_input.ts:
```diff
@@ -0,0 +1,47 @@
+import {applyValueToInputField} from '../apply_value_input_field';
+import {DirectiveDef, InputFlags} from '../interfaces/definition';
+
+export function writeToDirectiveInput<T>(
```
super-nit: this is not enforced / followed today but the initial idea was that the `instructions` folder would be just the entry point to the instructions called from the generated code with the actual logic living in the main `renderer3` folder (in the, say `inputs.ts`). 

Curious about your take on it.

devversion replied:
That makes sense. Yeah, I think it's not fully compliant right now, but I agree. Will follow-up on that in separate PR.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: size-tracking
Reviewed-for: fw-core

devversion committed:
fix(core): do not accidentally inherit input transforms when overridden

Currently when a base class defines an input with a transform, derived
classes re-defining the input via `@Input`, or `inputs: [<..>]`, end up
inherting the transform due to a bug in the inherit definitions feature.

This commit fixes this. We verified in the Google codebase that this is
an unlikely occurrence and it's trivial to fix on user side by removing
the re-declaration/override, or explictly adding the necessary
transform.

Conceptually, the behavior was quite inconsistent as everything else of
inputs was overridden as expected. i.e. alias, required state etc. The
exception were input transforms. This commit fixes this.

jessicajaniuk reviewed (approved):
reviewed-for: public-api

devversion commented:
@jessicajaniuk Thanks for the review. I think we can point them to the two small commits. Those should be fairly easy to review if they are interested: https://github.com/angular/angular/pull/53571/commits/1bcbd25300e3cba6382f0e32eae8974c514388d0, https://github.com/angular/angular/pull/53571/commits/44bca0a5546708348bd8ae5cd7ffe570418b5485

devversion commented:
Passing [TGP](https://fusion2.corp.google.com/presubmit/592539876/OCL:592539876:BASE:592807486:1703234635448:277fd7af/targets)

devversion committed:
fixup! refactor(core): detect signal inputs at runtime using input flags

Account for rebase and new local compilation tests

devversion committed:
fixup! refactor(compiler-cli): reference `InputFlags` enum directly for full compiler output

Account for rebase and new local compilation tests

atscott commented:
This PR was merged into the repository by commit 36318dbd41dbb9ffb1b5e45c226f4b6395834319.
